### PR TITLE
Implement certificate checks

### PR DESCRIPTION
**Pull request by Apollon77**
This PR implements all required Certificate checks and enahnces the Tlv -> ASN/X509 conversion to comply with the specs

Can be reviewed commit wise.

Includes also:
* Verifies Certificates on commissioning and also on Case session esteblishment
* Enhanced DER encoding to add missing types (GeneralizedTime, PrinableString, IA5String, Long numbers as BitStrings, ...)
* Handles Tlv Lists more near to the spec by respecting the key order in the object rather then the field definition - this is then utilized by conversion for subject, issuer and extension
* Corrects the encoding of the certificate serialnumber to allow more then 255 certificates
* Enhanced certificate testing by utilizing productions certs from Apple, Google, Amazon and SmartThings (where each ecosystem do something different)
* laxes checks for Key-Usage field comparedto specs because Amazon and Smartthings use additional flags
* Restructures Certificate Tlv/Type definitions and reduce duplicate code

Testing wise the new enhanced certificate testing (CertificateManagerTest), mainly because each ecosystem does something different, includes the following:
* Combinations of different fields in issuer and subject and different orders of these fields
* Google uses additional fields in subject (commonName e.g.)
* Google uses a PrintedString encoded field countryName in subject
* Apple uses the non-well defined data for not-after field
* Amazon/SmartThings use additional values in keyUsage BitString
* Our own Root certificateManager generates also certs with a uniqne non-tag-id-sorted order of subject/issuer fields (this is then tested via IntegrationTest)

### CODE REVIEW

File: packages/matter.js/src/util/Number.ts [hunk 1]
```diff
@@ -42,6 +42,17 @@ export function maxValue<T extends bigint | number>(a: T | undefined, b: T | und
     return a > b ? a : b;
 }
 
-export function toHexString(value: number | bigint) {
+/**
+ * Convert a number or bigint to a hex string which is prefixed by "0x" for logging purposes
+ */
+export function toLogHexString(value: number | bigint) {
```
Comment (lauckhart) [inline packages/matter.js/src/util/Number.ts:48]: An alternate place to put this could be Diagnostic.hex()

**Review by lauckhart — APPROVED** [review]
Wow a thing of beauty.  Glad it was you and not me 😁

File: packages/matter.js/src/behavior/definitions/operational-credentials/OperationalCredentialsServer.ts [hunk 2]
```diff
@@ -175,9 +167,22 @@ export class OperationalCredentialsServer extends OperationalCredentialsBehavior
...
+            } else if (
+                error instanceof CertificateError ||
+                error instanceof ValidationError ||
+                error instanceof UnexpectedDataError
+            ) {
+                return {
+                    statusCode: OperationalCredentials.NodeOperationalCertStatus.InvalidNoc,
+                    debugText: error.message,
+                };
+            } else if (error instanceof PublicKeyError) {
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/behavior/definitions/operational-credentials/OperationalCredentialsServer.ts:179]: But in fact they are no StatusReponseErrors ... they are here converted into a "real response" ... But correct - see below ... we can streamline the code and remove duplication

File: packages/matter.js/src/behavior/definitions/operational-credentials/OperationalCredentialsServer.ts [hunk 3]
```diff
@@ -331,7 +356,19 @@ export class OperationalCredentialsServer extends OperationalCredentialsBehavior
...
+        try {
+            failsafeContext.setRootCert(rootCaCertificate);
+        } catch (error) {
+            logger.info("setting root certificate failed", error);
+            if (
+                error instanceof CertificateError ||
+                error instanceof ValidationError ||
+                error instanceof UnexpectedDataError
+            ) {
+                throw new StatusResponseError(error.message, StatusCode.InvalidCommand);
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/behavior/definitions/operational-credentials/OperationalCredentialsServer.ts:368]: In fact this is the only place where it gets into a StatusResponseError and UnexpectedData/Validation are more common cases that are not always returned as InvalidCommand but velidatio mostly ConstraintError ... I reduced comde duplication a bit

File: packages/matter.js/src/certificate/CertificateManager.ts [hunk 4]
```diff
@@ -33,13 +24,47 @@ import { CaseAuthenticatedTag, TlvCaseAuthenticatedTag } from "../datatype/CaseA
...
+    Pkcs7SignedData,
+    Pseudonym_X520,
+    PublicKeyEcPrime256v1_X962,
+    SerialNumber_X520,
+    SHA256_CMS,
+    StateOrProvinceName_X520,
+    SubjectKeyIdentifier_X509,
+    SurName_X520,
+    Title_X520,
+} from "./CertificateDerTypes.js";
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/certificate/CertificateManager.ts:65]: good idea ...

File: packages/matter.js/src/cluster/server/OperationalCredentialsServer.ts [hunk 5]
```diff
@@ -142,6 +132,7 @@ export const OperationalCredentialsClusterHandler: (
                 caseAdminSubject,
             });
         } catch (error) {
+            logger.info("building fabric failed", error);
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/cluster/server/OperationalCredentialsServer.ts:135]: Intended. I adjusted the text

File: packages/matter.js/src/codec/DerCodec.ts [hunk 6]
```diff
@@ -139,18 +192,46 @@ export class DerCodec {
...
+        }
+        return this.encodeAnsi1(DerType.PrintableString, ByteArray.fromString(value));
+    }
+
+    private static encodeIA5String(value: string) {
+        /*eslint-disable-next-line no-control-regex */
+        if (!/^[\x00-\x7F]*$/.test(value)) {
+            throw new UnexpectedDataError(`String ${value} is not an IA5 string.`);
+        }
+        return this.encodeAnsi1(DerType.IA5String, ByteArray.fromString(value));
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/codec/DerCodec.ts:207]: Ha, right ... fixed

File: packages/matter.js/src/util/Number.ts (hunk 1 above)
Comment (Apollon77) [reply packages/matter.js/src/util/Number.ts]: Please have a look ... I did it as normal functiopn in the namesace ... you meaned it that way, right?

File: packages/matter.js/src/codec/DerCodec.ts [hunk 7]
```diff
@@ -68,89 +83,155 @@ export class DerCodec {
...
+                        .toISOString()
+                        .replace(/[-:.T]/g, "")
+                        .slice(2, 14) + "Z",
+                ),
+            );
     }
 
     private static encodeBoolean(bool: boolean) {
-        return this.encodeAnsi1(DerType.Boolean, ByteArray.of(bool ? 0xff : 0));
+        return this.encodeAns1(DerType.Boolean, ByteArray.of(bool ? 0xff : 0));
```
Comment (lauckhart) [inline packages/matter.js/src/codec/DerCodec.ts:169]: Need to swap "n" and "s" too 🙂... Am on mobile at the moment or I'd do it

**Review by lauckhart — APPROVED** [review]

File: packages/matter.js/src/util/Number.ts (hunk 1 above)
Comment (lauckhart) [reply packages/matter.js/src/util/Number.ts]: Yep that's right

File: packages/matter.js/src/codec/DerCodec.ts (hunk 7 above)
Comment (Apollon77) [reply packages/matter.js/src/codec/DerCodec.ts:169]: Oohhh. Lol. Missed that. I do

**Pull request merged**